internal/pkg/dns/dns.go

```diff
@@ -0,0 +1,251 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package dns provides dns server implementation.
+package dns
+
+import (
+	"context"
+	"errors"
+	"net"
+	"net/netip"
+	"sync"
+	"time"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/cache"
+	"github.com/coredns/coredns/plugin/pkg/proxy"
+	"github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
+	"github.com/siderolabs/gen/pair"
+	"github.com/siderolabs/gen/xslices"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/talos/internal/pkg/utils"
+)
+
+// NewRunner creates a new Runner.
+func NewRunner(srv Server, logger *zap.Logger) *Runner {
+	return &Runner{srv: srv, logger: logger, done: make(chan struct{})}
+}
+
+// Runner is a dns server handler.
+type Runner struct {
+	srv    Server
+	logger *zap.Logger
+	done   chan struct{}
+}
+
+// Server is a dns server.
+type Server interface {
+	ListenAndServe() error
+	Shutdown() error
+}
+
+// Run runs dns server.
+func (h *Runner) Run() error {
+	defer close(h.done)
+
+	h.logger.Info("starting dns server")
+
+	err := h.srv.ListenAndServe()
+
+	h.logger.Info("dns server stopped", zap.Error(err))
+
+	return err
+}
+
+// Stop stops dns server. It's safe to call even if server is already stopped. It's not safe to call from multiple
+// goroutines.
+func (h *Runner) Stop() {
+	select {
+	case <-h.done:
+		return
+	default:
+	}
+
+	err := h.srv.Shutdown()
+	if err != nil {
+		h.logger.Warn("error shutting down dns server", zap.Error(err))
+	}
+
+	<-h.done
+}
+
+// Cache is a [dns.Handler] to [plugin.Handler] adapter.
+type Cache struct {
+	cache  *cache.Cache
+	logger *zap.Logger
+}
+
+// NewCache creates a new Cache.
+func NewCache(next plugin.Handler, l *zap.Logger) *Cache {
+	c := cache.NewCache("zones", "view")
+	c.Next = next
+
+	return &Cache{cache: c, logger: l}
+}
+
+// ServeDNS implements [dns.Handler].
+func (c *Cache) ServeDNS(wr dns.ResponseWriter, msg *dns.Msg) {
+	_, err := c.cache.ServeDNS(context.Background(), wr, msg)
+	if err != nil {
+		// we should probably call newProxy.Healthcheck() if there are too many errors
+		c.logger.Warn("error serving dns request", zap.Error(err))
+	}
+}
+
+// Handler is a dns proxy selector.
+type Handler struct {
+	mx     sync.RWMutex
+	dests  []pair.Pair[netip.Addr, *proxy.Proxy]
+	logger *zap.Logger
+}
+
+// NewHandler creates a new Handler.
+func NewHandler(logger *zap.Logger) *Handler {
+	return &Handler{
+		logger: logger,
+	}
+}
+
+// Name implements plugin.Handler.
+func (h *Handler) Name() string {
+	return "Handler"
+}
+
+// ServeDNS implements plugin.Handler.
+func (h *Handler) ServeDNS(ctx context.Context, wrt dns.ResponseWriter, msg *dns.Msg) (int, error) {
+	h.mx.RLock()
+	defer h.mx.RUnlock()
+
+	req := request.Request{W: wrt, Req: msg}
+
+	var (
+		resp *dns.Msg
+		err  error
+	)
+
+	h.logger.Info("dns request", zap.Stringer("data", msg))
+
+	for {
+		prx, ok := randomGet(h.dests)
```

Comment:
We should probably try all upstreams before we give up (CoreDNS actually does this).